File: RetroGameGauntlet.Droid/Services/PlatformLoaderService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Globalization;
using RetroGameGauntlet.PCL.Services;
using RetroGameGauntlet.PCL.Models;
using System.Threading.Tasks;

namespace RetroGameGauntlet.Droid.Services
{
    public class PlatformLoaderService : IPlatformLoaderService
    {
        private List<KeyValuePair<string, string>> platforms;

        #region IPlatformLoader implementation

        public async Task<string> GetRandomGameAsync(string platform)
        {
            string rvalue = null;
            await Task.Run(() =>
            {
                var fileReader = new StreamReader(Xamarin.Forms.Forms.Context.Assets.Open("platforms/" + platform));
                string nextLine;
                List<string> games = new List<string>();
                while ((nextLine = fileReader.ReadLine()) != null)
                {
                    games.Add(nextLine);
                }
                fileReader.Close();

                Random rnd = new Random();
                rvalue = games[rnd.Next(0, games.Count)];
            });
            return rvalue;
        }


        public async Task<IEnumerable<KeyValuePair<string, string>>> FindGamesForAsync(string query)
        {
            if (platforms == null)
            {
                platforms = Platforms.All.Select((arg) => new KeyValuePair<string, string>(arg.FileName, arg.Title)).ToList();
            }
            List<KeyValuePair<string, string>> games = new List<KeyValuePair<string, string>>();
            await Task.Run(() =>
            {
                foreach (var platform in platforms)
                {
                    var fileReader = new StreamReader(Xamarin.Forms.Forms.Context.Assets.Open("platforms/" + platform.Key));
                    string nextLine;
                    while ((nextLine = fileReader.ReadLine()) != null)
                    {
                        if (CultureInfo.CurrentCulture.CompareInfo.IndexOf(nextLine, query, CompareOptions.IgnoreCase) >= 0)
                        {
                            games.Add(new KeyValuePair<string, string>(nextLine, platform.Value));
                        }
                    }
                    fileReader.Close();
                }
            });
            return games;
        }

        #endregion
    }
}

File: RetroGameGauntlet.Droid/Services/PlatformLoaderService.fs
namespace RetroGameGauntlet.Droid.Services

open System
open System.Collections.Generic
open System.Globalization
open System.IO
open System.Linq
open Plugin.CurrentActivity
open RetroGameGauntlet.Core.Models
open RetroGameGauntlet.Core.Services

type PlatformLoaderService () =
    let platforms = Platforms.All |> Seq.map(fun (arg) -> KeyValuePair<string, string>(arg.FileName, arg.Title))
    let includes(a:string, b:string) = CultureInfo.InvariantCulture.CompareInfo.IndexOf(a, b, CompareOptions.IgnoreCase) >= 0
    let activity() = CrossCurrentActivity.Current.Activity

    interface IPlatformLoaderService with
        /// Gets the random game for a platform.
        member this.GetRandomGameAsync platform =
            async {
                let filePath = activity().Assets.Open("platforms/" + platform)
                let fileReader = new StreamReader(filePath)
                let games = List<string>()
                let mutable nextLine: string = fileReader.ReadLine()
                while nextLine <> null do
                    games.Add nextLine
                    nextLine <- fileReader.ReadLine()
                fileReader.Close()
                let rnd = Random()
                return games.Item(rnd.Next(0, games.Count))
            }

        /// Gets the full list of games for a platform.
        member this.FindGamesForAsync query =
            async {
                let games = List<GameModel>()
                platforms 
                    |> Seq.iter(
                        fun platform -> 
                            let filePath = activity().Assets.Open("platforms/" + platform.Key)
                            let fileReader = new StreamReader(filePath)
                            let mutable nextLine: string = fileReader.ReadLine()
                            while nextLine <> null do
                                if includes(nextLine, query) then
                                    games.Add(GameModel(nextLine, platform.Value))
                                nextLine <- fileReader.ReadLine()
                            fileReader.Close()
                    )
                return games.AsEnumerable()
            }